lsp-methods.el

```diff
@@ -707,25 +708,50 @@ interface Range {
   "Return the end line for a given LSP range, in LSP coordinates"
   (inline-quote (plist-get (plist-get ,range :end) :line)))
 
-(defun lsp--apply-workspace-edits (edits)
-  (cl-letf (((lsp--workspace-change-timer-disabled lsp--cur-workspace) t))
-    (maphash (lambda (key value)
-               (lsp--apply-workspace-edit key value))
-             (gethash "changes" edits))))
-
-(defun lsp--apply-workspace-edit (uri edits)
-  (let ((filename (string-remove-prefix lsp--uri-file-prefix uri)))
-    ;; TODO: What if the buffer has been modified?
-    ;;       Although, for incremental sync that should be fine
-    (when (not (find-buffer-visiting filename))
-      (progn (find-file filename)
-             (lsp--text-document-did-open)))
-    (lsp--apply-text-edits edits)))
+(defun lsp--apply-workspace-edit (edit)
+  "Apply the WorkspaceEdit object EDIT.
 
-(defun lsp--apply-text-edits (edits)
+interface WorkspaceEdit {
+	changes?: { [uri: string]: TextEdit[]; };
+	documentChanges?: TextDocumentEdit[];
+}"
+  (let ((changes (gethash "changes" edit))
+         (document-changes (gethash "documentChanges" edit)))
+    (if document-changes
+      (mapc #'lsp--apply-text-document-edit document-changes)
+
+      (maphash
```

Comment:
When documentChanges does not exist and changes is an empty list, `maphash` will error on `nil` input.